engines/e_afalg.c

```diff
@@ -421,6 +421,7 @@ static int afalg_start_cipher_sk(afalg_ctx *actx, const unsigned char *in,
 # endif
     char cbuf[CMSG_SPACE(ALG_IV_LEN(ALG_AES_IV_LEN)) + CMSG_SPACE(ALG_OP_LEN)];
 
+    memset(&msg, 0, sizeof(msg));
```

Comment:
What was it that caused this change?

Reply:
this failure in the enable-msan build, but it is not fixed by this: 
```
Uninitialized bytes in read_msghdr_control at offset 4 inside [0x7ffcea486118, 20)
==29894==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x4399cc in __interceptor_sendmsg /tmp/final/llvm.src/projects/compiler-rt/lib/msan/../sanitizer_common/sanitizer_common_interceptors.inc:3030:5
    #1 0x7fc4944fd127 in afalg_start_cipher_sk /home/travis/build/openssl/openssl/engines/e_afalg.c:483:14
    #2 0x7fc4944fd127 in afalg_do_cipher /home/travis/build/openssl/openssl/engines/e_afalg.c:601
    #3 0x7fc496c746f1 in EVP_EncryptFinal_ex /home/travis/build/openssl/openssl/crypto/evp/evp_enc.c:435:11
    #4 0x493cbe in test_afalg_aes_cbc /home/travis/build/openssl/openssl/test/afalgtest.c:91:97
    #5 0x495bb4 in run_tests /home/travis/build/openssl/openssl/test/testutil/driver.c:390:23
    #6 0x496799 in main /home/travis/build/openssl/openssl/test/testutil/main.c:30:15
    #7 0x7fc495af382f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)
    #8 0x41c8a8 in _start (/home/travis/build/openssl/openssl/test/afalgtest+0x41c8a8)

SUMMARY: MemorySanitizer: use-of-uninitialized-value /tmp/final/llvm.src/projects/compiler-rt/lib/msan/../sanitizer_common/sanitizer_common_interceptors.inc:3030:5 in __interceptor_sendmsg
Exiting
../../util/shlib_wrap.sh ../../test/afalgtest => 77
not ok 1 - running afalgtest
```

Reply:
Huh?  I thought that `struct msghdr msg = { 0 };` was supposed to initialize to zero properly...

Reply:
Me too.
This is probably the wrong way to fix, maybe the msan see somthing in the msghdr
pointing to uninitialized memory?

Reply:
Well, it's not exactly *wrong*, a `memset` to zero is just as appropriate, and according to some texts I just read, gets the point across more clearly...  so I don't mind this change, I was just a bit surprised.

Reply:
But wait, `sendmsg` gets more parameters...  could it be that `actx->sfd` isn't properly initialized?  It's a wild guess, I need to study the code to know for sure

Reply:
Ah, no, that isn't it...

Reply:
```
Uninitialized bytes in read_msghdr_control at offset 4 inside [0x7ffcea486118, 20)
```

this could indicate that whatever `msg.msg_control` points at isn't quite initialized...  but we *do* initialize `cbuf`, so that's weird.

Still guessing wildly

Reply:
Another wild guess, could it be an alignment issue, i.e. that `msg.msg_control` must be a properly aligned pointer, i.e. it must point at `cbuf + alignmentoffset` rather than `cbuf` directly (`alignmentoffset` might or might not be zero)?

Still guessing wildly

Reply:
It could be that we copy someting uninitialized into cbuf at offset 4.

Reply:
The only place I see us doing anything with `cbuf` is this:

``` C
    memset(cbuf, 0, sizeof(cbuf));
```

Reply:
but cbuf starts with struct cmsghdr::cmsg_len which is of type size_t.
weird...

Reply:
No.  `cbuf` is a `char` array, so its alignment will be whatever fits after the previous local variable, which so happens to be an `int` if `ALG_ZERO_COPY` is defined.

Reply:
`4 inside [0x7ffcea486118, 20)` so cmsg is 8-byte aligned. but only by chance.
what I don't understand, is 20=4*5, so the size is not a multiple of 8, but sizeof(size_t) is 8.
CMSG_SPACE should round up to multiple of 8.
or is that hex 20?

Reply:
`CMSG_SPACE(ALG_IV_LEN(ALG_AES_IV_LEN)) + CMSG_SPACE(ALG_OP_LEN)`

I can imagine that 20 = 2*8 + 4...  althougt I'm not sure I quite understand the numbers, 'cause `ALG_OP_LEN` is `sizeof(unsigned int)` and `ALG_IV_LEN(ALG_AES_IV_LEN)` is `sizeof(struct af_alg_iv) + 16`, so that doesn't quite add up...  `CMSG_SPACE` is supposed to ensure enough space for alignment...

Reply:
Yes, the possible mis-alignment of the buffer is probably a real issue.
But changing it to size_t cbuf[..] does not help at all.
even 0x20 is impossible, the cbuf must have space for at least 2xsize_t
and 16 byte IV, that alone is 0x20...

Reply:
I lean towards disable-afalgeng in this test.